### PR TITLE
Embed comit-i in comit_node

### LINKED ISSUE #855: Embed comit-i in comit-rs
Follow-up of coblox/comit-i#17

DoD:
- [x] Research and how exactly to do it
- [x] Discuss release strategy
- [x] Implement release strategy

### PR DESCRIPTION
**Pull request by D4nte**
Resolves #855

### CODE REVIEW

File: api_tests/regtest/alice/default.toml [hunk 1]
```diff
@@ -5,10 +5,13 @@ secret_seed = "f87165e305b0f7c4824d3806434f9d0909610a25641ab8773cf92a48c9d77670"
 [http_api]
 address="0.0.0.0"
 port=8000
-logging=true
+
+[comit_i]
```
[the remark this answers is not in the document]
Comment (luckysori) [reply api_tests/regtest/alice/default.toml]: Or maybe add it to Bob's and Charlie's configs if you want these to be complete config files.

File: vendor/comit_i/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[package]
+authors = ["Franck Royer <franck@coblox.tech>"]
```
Comment (luckysori) [inline vendor/comit_i/Cargo.toml]: Rude.

File: api_tests/regtest/alice/default.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply api_tests/regtest/alice/default.toml]: 1. We do for the cross-origin auth
2. It's actually good to have it, it allows you to quickly start a node with comit-i using `COMIT_NODE_CONFIG_PATH=./api_tests/regtest/alice/ cargo run --package comit_node`

File: vendor/comit_i/Cargo.toml (hunk 2 above)
Comment (D4nte) [reply vendor/comit_i/Cargo.toml]: Do `cargo init` locally and you'll see.

File: api_tests/regtest/alice/default.toml (hunk 1 above)
Comment (thomaseizinger) [inline api_tests/regtest/alice/default.toml]: :wrench: Should we just call this `web_gui`?

File: vendor/comit_i/Cargo.toml (hunk 2 above)
Comment (luckysori) [reply vendor/comit_i/Cargo.toml]: Of course you'd want to claim comit-i yours.

File: application/comit_node/src/comit_i_routes.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            Error::PathConversionFail => "Conversion of the path to PathAndQuery failed",
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
+        None
+    }
+}
+
+pub fn create() -> BoxedFilter<(impl Reply,)> {
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_i_routes.rs]: :wrench: I liked it better when this was part of the `comit_i` crate. Why did you remove it?

File: application/comit_node/src/comit_i_routes.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            Asset::get("index.html")
+        })
+        .ok_or_else(|| Error::IndexHtmlMissing)?;
+
+    Ok(Response::builder()
+        .header("content-type", mime.to_string())
+        .body(file))
+}
+
+pub fn unpack_problem(rejection: Rejection) -> Result<impl Reply, Rejection> {
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_i_routes.rs]: :confused: This is not unpacking a `problem` (you are not using HttpApiProblem) - so I think the function name should not indicate that.

File: application/comit_node/src/comit_i_routes.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    if let Some(&err) = rejection.find_cause::<Error>() {
+        let code = match err {
+            Error::IndexHtmlMissing => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::PathConversionFail => StatusCode::BAD_REQUEST,
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_i_routes.rs]: :question: Why is a BAD REQUEST? Sounds like an internal error to me.

File: api_tests/regtest/alice/default.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (luckysori) [reply api_tests/regtest/alice/default.toml]: I see now! I like that then. Please add it to the other configs though :wink:

File: application/comit_node/src/comit_i_routes.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        .boxed()
+}
+
+fn serve(path: Tail) -> Result<impl Reply, Rejection> {
+    let path = path.as_str();
+    let path_and_query = PathAndQuery::from_str(path).map_err(|e| {
+        error!("Could not convert path {} to PathAndQuery: {:?}", path, e);
+        Error::PathConversionFail
+    })?;
+    let path = path_and_query.path();
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_i_routes.rs]: :confused: How many times do you want to convert back and forth between `Path`s? :see_no_evil: 

Is there no easier way to do this?

The function `guess_mime_type_opt` [takes anything that is `AsRef<Path>`](https://docs.rs/mime_guess/2.0.0-alpha.6/mime_guess/fn.guess_mime_type_opt.html), which is implemented for `&str` if I remember correctly. You should be able to simply pass `path.as_str()` to it.

Also, it looks like using [`warp::path::full()`](https://docs.rs/warp/0.1.15/warp/filters/path/struct.FullPath.html) might be more appropriate than `warp::path::tail()` because you actually want to match against the full thing and not the remaining, unmatched part.
In your concrete case, it is semantically equivalent because you don't do any matching before.

File: application/comit_node/src/comit_i_routes.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    let mut mime = mime_guess::guess_mime_type_opt(path)
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_i_routes.rs]: :wrench: Maybe there is a way to avoid the mutability + double parsing here:

1. `let asset: Option<Cow<'static, [u8]>> = Asset::get(path);`
2. `map` asset to a response + guess the mimetype ([`guess_mime_type`](https://docs.rs/mime_guess/2.0.0-alpha.6/mime_guess/fn.guess_mime_type.html) seems to more appropriate here instead of defaulting to `text/html`)
3. `unwrap_or_else` the option to serve the `index.html` asset by constructing a `Response` that statically sets the `content-type` header to "text/html" (atm you are parsing first and a `to_string` afterwards.

You can fetch the `index.html` asset upfront, since nothing works without that one anyway :)

```rust
let index = Asset::get("index.html").ok_or_else(|| Error::IndexHtmlMissing)?
```

Comment (thomaseizinger) [reply application/comit_node/src/comit_i_routes.rs]: This causes a bit of duplication in constructing the `Response`, however, I'd argue that it is overall preferable since you keep the special-case of serving the `index.html` in case the asset is not found separate.

You could even extract it in its own function to not clutter up this one.

File: vendor/comit_i/build.rs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+use std::fs::File;
+
+fn main() {
+    let version = include_str!("./version").trim();
```
Comment (thomaseizinger) [inline vendor/comit_i/build.rs]: It might be useful to do some logging here like
- which version of comit-i we are downloading
- where the tmp file is located

Reference: https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script

File: application/comit_node/src/comit_i_routes.rs (hunk 5 above)
Comment (LLFourn) [reply application/comit_node/src/comit_i_routes.rs]: I guess because it means the client has requested something that doesn't exist (it should probably be 404).

**Comment by LLFourn** [comment]
This seems to need something in gitignore:

```
On branch 855-embed-comit-i
Your branch is up to date with 'origin/855-embed-comit-i'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	vendor/comit_i/build/

nothing added to commit but untracked files present (use "git add" to track)
```

File: application/comit_node/src/comit_i_routes.rs (hunk 4 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: I must say I find the handling of error with warp extremely confusing. Feel free to 🔧 something cleaner/simpler.

File: application/comit_node/src/comit_i_routes.rs (hunk 3 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: What was in the comit_i crate was just an example to use rust-embed with warp. I have now moved it directly to rust-embed repo.

To me, the comit_i crate just provides comit_i. The lib consumer should have the choice on how to serve it (warp, rocket, etc).
I did not want to make this choice for the consumer that is why it is in comit_node crate.

File: application/comit_node/src/comit_i_routes.rs (hunk 5 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: this path is coming from the client and if the conversion to a "path" fails it means it's a bad request.

File: application/comit_node/src/comit_i_routes.rs (hunk 6 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: > 😕 How many times do you want to convert back and forth between Paths? 🙈
I am sure you can find a better way to phrase your point.

I'll tackle the rest of the comment.

Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: Please note that unfortunately warp hides the "PathAndQuery" from the `Tail` struct making this cumbersome.

I'll see if I can make it better with `warp::path::full()`

File: application/comit_node/src/comit_i_routes.rs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        .unwrap_or_else(|| "text/html".parse::<Mime>().unwrap());
```
[the remark this answers is not in the document]
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: haha no I cannot. The `mime` is version `0.3` whereas `mine_guess` internally uses `0.2`. It re-exports the `Mime` struct from the `mime` crate but that's it. Hence I cannot use the correct `mime::TEXT_HTML`.

I'll do a PR on `mime_guess`.

**Comment by D4nte** [comment]
> This seems to need something in gitignore:

Weird, I'll see if I can correct but the `.gitignore` is there.

File: application/comit_node/src/comit_i_routes.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: https://github.com/abonander/mime_guess/pull/44

File: api_tests/regtest/alice/default.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (LLFourn) [reply api_tests/regtest/alice/default.toml]: After discussion can we:

1. Remove this from here
2. Have COMIT-i tests in its own directory and have a `config.toml` that sets the config dir to something specific to COMIT-i tests (i.e. not `./api_tests/regtest/alice/`)

To make the above work you'll need to change the harness so it doesn't run tests from `api_tests` as it's cwd. This will cause some collateral damage in various places.

File: application/comit_node/src/comit_i_routes.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LLFourn) [reply application/comit_node/src/comit_i_routes.rs]: Heh. You can still do it better in the old version:
`mime::mime!(Text/Html)`

There's an example in the docs:
```rust
#[macro_use] extern crate mime;
fn main() {
     let json = mime!(Application/Json);
     let plain = mime!(Text/Plain; Charset=Utf8);
     let text = mime!(Text/Html; Charset=("bar"), ("baz")=("quux"));
     let img = mime!(Image/_);
}
```

File: application/comit_node/src/comit_i_routes.rs (hunk 5 above)
Comment (LLFourn) [reply application/comit_node/src/comit_i_routes.rs]: Ah good point. If the file doesn't exist what happens?

File: application/comit_node/src/comit_i_routes.rs (hunk 4 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: sorted now

File: application/comit_node/src/comit_i_routes.rs (hunk 5 above)
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: This has been removed :)

File: application/comit_node/src/comit_i_routes.rs (hunk 3 above)
Comment (thomaseizinger) [reply application/comit_node/src/comit_i_routes.rs]: Hmm, comit_i by itself is provided in its own repository. Everything that is in the `comit_i` crate can be trivially built by everyone (no real gain of re-use here).

For now though, only the functionality matters :)

File: application/comit_node/src/comit_i_routes.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply application/comit_node/src/comit_i_routes.rs]: I don't have access to the correct version of `mime`. Code has been removed anyway.

File: vendor/comit_i/Cargo.toml (hunk 2 above)
Comment (D4nte) [reply vendor/comit_i/Cargo.toml]: God no!

File: api_tests/regtest/alice/default.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply api_tests/regtest/alice/default.toml]: The 🚀 is here to show that the scope creep is heading to the moon.

File: vendor/comit_i/build.rs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+    let url = format!(
+        "https://github.com/comit-network/comit-i/releases/download/{}/bundle.zip",
+        version
+    );
+
+    println!("Fetching {}", url);
+    let mut tmpfile: File = tempfile::tempfile().unwrap();
+    reqwest::get(&url).unwrap().copy_to(&mut tmpfile).unwrap();
+    unzip::Unzipper::new(tmpfile, "./").unzip().unwrap();
```
Comment (LLFourn) [inline vendor/comit_i/build.rs]: 🔧  Use expect for these.
